### PR TITLE
[runtime] Fix C and LLVM atomics on ARM64

### PR DESCRIPTION
**Pull request by alexrp**

### CODE REVIEW

**Comment by alexrp** [comment]
Please don't merge unless it's actually green on ARM64.

**Comment by alexrp** [comment]
@vargaz does the LLVM bit look ok?

**Comment by vargaz** [comment]
Yes.

**Comment by alexrp** [comment]
cc @kumpera @BrzVlad

File: mono/utils/atomic.h [hunk 1]
```diff
@@ -180,38 +180,71 @@ static inline void InterlockedWrite16(volatile gint16 *dst, gint16 val)
...
+ * intrinsic if we detect that we're built with a buggy compiler.
+ */
+
+#if defined (HOST_ARM64) && (defined (__clang__) || MONO_GNUC_VERSION < 50300)
+#define WRAP_ATOMIC_INTRINSIC(RET, INTRIN) \
+	({ \
+		mono_memory_barrier (); \
+		__typeof__ (INTRIN) atomic_ret_ ## RET ## __ = INTRIN; \
+		mono_memory_barrier (); \
+		atomic_ret_ ## RET ## __; \
```
Comment (BrzVlad) [inline mono/utils/atomic.h]: Do we really need unique names for this variable ?

Comment (kumpera) [inline mono/utils/atomic.h]: Given the var is scoped, I wonder the same. It's a style issue, though.

**Comment by kumpera** [comment]
LGTM

Good catch on removing ia64 dead code and unifying gcc version numbering.

There's the style issue that Vlad pointed out, but that doesn't invalidate this PR itself.

File: mono/utils/atomic.h (hunk 1 above)
Comment (alexrp) [inline mono/utils/atomic.h]: Probably a bit overengineered. Will simplify before merging.

**Pull request merged**